src/v0/destinations/fb_custom_audience/transformV2.js

```diff
@@ -0,0 +1,249 @@
+
+        dataElement = getUpdatedDataElement(
+          dataElement,
+          isHashRequired,
+          eachProperty,
+          updatedProperty,
+        );
+
+        if (dataElement[dataElement.length - 1]) {
+          nullUserData = false;
+        }
+      });
+
+      if (nullUserData) {
+        stats.increment('fb_custom_audience_event_having_all_null_field_values_for_a_user', {
+          destinationId: destination.ID,
+          nullFields: userSchema,
+        });
+      }
+
+      data.push(dataElement);
+    });
+
+    const prepareFinalPayload = lodash.cloneDeep(paramsPayload);
+    prepareFinalPayload.schema = userSchema;
+    prepareFinalPayload.data = data;
+    const payloadBatches = batchingWithPayloadSize(prepareFinalPayload);
+
+    payloadBatches.forEach((payloadBatch) => {
+      const response = {
+        ...prepareParams,
+        payload: payloadBatch,
+      };
+      const wrappedResponse = {
+        responseField: response,
+        operationCategory: operation,
+      };
+      toSendEvents.push(wrappedResponse);
+    });
+  });
+
+  return toSendEvents;
+};
+
+async function processRecordInputs(groupedRecordInputs) {
+  const { destination } = groupedRecordInputs[0];
+  const { message } = groupedRecordInputs[0];
+  const { isHashRequired, accessToken, disableFormat, type, subType, isRaw, maxUserCount } =
+    destination.Config;
+  const prepareParams = {
+    access_token: accessToken,
+  };
+
+  // maxUserCount validation
+  const maxUserCountNumber = parseInt(maxUserCount, 10);
+  if (Number.isNaN(maxUserCountNumber)) {
+    throw new ConfigurationError('Batch size must be an Integer.');
+  }
+
+  // audience id validation
+  const { audienceId } = destination.Config;
+  let operationAudienceId = audienceId;
+  const mappedToDestination = get(message, MappedToDestinationKey);
+  if (!operationAudienceId && mappedToDestination) {
+    const { objectType } = getDestinationExternalIDInfoForRetl(message, 'FB_CUSTOM_AUDIENCE');
+    operationAudienceId = objectType;
+  }
+  if (!isDefinedAndNotNullAndNotEmpty(operationAudienceId)) {
+    throw new ConfigurationError('Audience ID is a mandatory field');
+  }
+
+  // user schema validation
+  let { userSchema } = destination.Config;
+  if (mappedToDestination) {
+    userSchema = getSchemaForEventMappedToDest(message);
+  }
+  if (!Array.isArray(userSchema)) {
+    userSchema = [userSchema];
+  }
+  if (!checkSubsetOfArray(schemaFields, userSchema)) {
+    throw new ConfigurationError('One or more of the schema fields are not supported');
+  }
+
+  const paramsPayload = {};
+  const dataSource = {};
+  if (isRaw) {
+    paramsPayload.is_raw = isRaw;
+  }
+  if (type && type !== 'NA' && typeFields.includes(type)) {
+    dataSource.type = type;
+  }
+  if (subType && subType !== 'NA' && subTypeFields.includes(subType)) {
+    dataSource.sub_type = subType;
+  }
+  if (Object.keys(dataSource).length > 0) {
+    paramsPayload.data_source = dataSource;
+  }
```

Comment:
This is repeated with audienceList type implementation can't we move this to utils and use it in both places?

Reply:
updated